File: pacc-cross-platform-updater/tools/pcu-patchgen/pcu_patchgen/bspatch.py
"""bsdiff 补丁的生成与应用（标准库实现）。

格式与 libbsdiff 一致：

    offset 0    8 字节  魔数 "BSDIFF40"
    offset 8    8 字节  CTRL 块压缩后的长度
    offset 16   8 字节  DIFF 块压缩后的长度
    offset 24   8 字节  新文件长度
    offset 32           CTRL / DIFF / EXTRA 三个独立 bzip2 流，依次排列

三个长度都是 bsdiff 的 off_t：小端 63 位量值，最高位当符号位。

CTRL 块是 (x, y, z) 三元组序列，语义是「从旧文件当前偏移复制 x 字节（逐字节加上
DIFF 块里的差值）/ 插入 y 字节（取自 EXTRA 块）/ 旧文件偏移再挪 z（可为负）」，
循环直到新文件装满。
"""

from __future__ import annotations

import bz2

MAGIC = b"BSDIFF40"
HEADER_SIZE = 32
TUPLE_SIZE = 24

# 生成端找匹配用的窗口长度。太短会在随机数据里撞出假匹配，太长则短重复段匹配不上。
WINDOW = 16


def _u64(value: int) -> bytes:
    return int(value).to_bytes(8, "little")


def read_offset(buf: bytes, off: int) -> int:
    """读一个 off_t（小端 63 位量值 + 最高位符号位）。"""
    if off + 8 > len(buf):
        raise ValueError("读取 off_t 越界")
    raw = int.from_bytes(buf[off:off + 8], "little")
    sign_bit = 1 << 63
    magnitude = raw & (sign_bit - 1)
    return -magnitude if raw & sign_bit else magnitude


def write_offset(value: int) -> bytes:
    """写一个 off_t。"""
    sign_bit = 1 << 63
    if value < 0:
        return _u64(sign_bit | (-value & (sign_bit - 1)))
    return _u64(value & (sign_bit - 1))


def _assemble(ctrl: bytes, diff: bytes, extra: bytes, new_size: int) -> bytes:
    ctrl_z = bz2.compress(ctrl)
    diff_z = bz2.compress(diff)
    extra_z = bz2.compress(extra)
    return (
        MAGIC
        + _u64(len(ctrl_z))
        + _u64(len(diff_z))
        + _u64(new_size)
        + ctrl_z
        + diff_z
        + extra_z
    )


def make_patch(old: bytes, new: bytes) -> bytes:
    """生成 ``old -> new`` 的 BSDIFF40 补丁（朴素贪心，保证确定性）。

    旧文件里所有长度为 WINDOW 的窗口进索引（同键保留首次出现位置），在 ``new`` 上
    从左往右扫，命中就向后尽可能延长匹配，落成「先插入待定段并 seek、再复制」两个
    三元组；没命中的字节积累成待插入段。
    """
    index: dict[bytes, int] = {}
    if len(old) >= WINDOW:
        for pos in range(0, len(old) - WINDOW + 1):
            index.setdefault(old[pos:pos + WINDOW], pos)

    ctrl = bytearray()
    diff = bytearray()
    extra = bytearray()
    old_pos = 0
    pending = 0  # 待插入段在 new 里的起点
    i = 0
    while i < len(new):
        pos = index.get(new[i:i + WINDOW]) if i + WINDOW <= len(new) else None
        if pos is None:
            i += 1
            continue
        mlen = WINDOW
        while (pos + mlen < len(old) and i + mlen < len(new)
               and old[pos + mlen] == new[i + mlen]):
            mlen += 1
        insert_len = i - pending
        if insert_len > 0 or pos != old_pos:
            ctrl += write_offset(0)
            ctrl += write_offset(insert_len)
            ctrl += write_offset(pos - old_pos)
            extra += new[pending:i]
            old_pos = pos
        ctrl += write_offset(mlen)
        ctrl += write_offset(0)
        ctrl += write_offset(0)
        for k in range(mlen):
            diff.append((new[i + k] - old[pos + k]) & 0xFF)
        i += mlen
        old_pos += mlen
        pending = i

    insert_len = len(new) - pending
    if insert_len > 0:
        ctrl += write_offset(0)
        ctrl += write_offset(insert_len)
        ctrl += write_offset(0)
        extra += new[pending:]
    return _assemble(bytes(ctrl), bytes(diff), bytes(extra), len(new))


def apply_patch(old: bytes, patch: bytes) -> bytes:
    """把补丁打到旧文件上，还原出新文件。

    校验从严：魔数不对、长度非法、三元组不够、复制或插入越过可用数据，都抛
    ``ValueError``，不静默截断也不补零。
    """
    if len(patch) < HEADER_SIZE:
        raise ValueError("补丁长度不足，装不下 BSDIFF40 头")
    if patch[:8] != MAGIC:
        raise ValueError("补丁魔数不是 BSDIFF40")

    ctrl_len = read_offset(patch, 8)
    diff_len = read_offset(patch, 16)
    new_size = read_offset(patch, 24)
    if ctrl_len < 0 or diff_len < 0 or new_size < 0:
        raise ValueError("补丁声明了负长度")
    if HEADER_SIZE + ctrl_len + diff_len > len(patch):
        raise ValueError("补丁声明的 CTRL/DIFF 长度超出实际数据")

    diff_start = HEADER_SIZE + ctrl_len
    extra_start = diff_start + diff_len
    ctrl = bz2.decompress(patch[HEADER_SIZE:diff_start])
    diff = bz2.decompress(patch[diff_start:extra_start])
    extra = bz2.decompress(patch[extra_start:])

    out = bytearray(new_size)
    old_pos = 0
    new_pos = 0
    ctrl_pos = 0
    diff_pos = 0
    extra_pos = 0
    while new_pos < new_size:
        if ctrl_pos + TUPLE_SIZE > len(ctrl):
            raise ValueError("CTRL 块不完整，缺少三元组")
        copy_len = read_offset(ctrl, ctrl_pos)
        insert_len = read_offset(ctrl, ctrl_pos + 8)
        seek_len = read_offset(ctrl, ctrl_pos + 16)
        ctrl_pos += TUPLE_SIZE
        if copy_len < 0 or insert_len < 0:
            raise ValueError("三元组出现负的复制/插入长度")
        if new_pos + copy_len > new_size:
            raise ValueError("复制长度超出新文件长度")
        if diff_pos + copy_len > len(diff):
            raise ValueError("DIFF 块数据不足")
        if copy_len > 0:
            if old_pos < 0 or old_pos + copy_len > len(old):
                raise ValueError("复制区间超出旧文件范围")
            for k in range(copy_len):
                out[new_pos + k] = (diff[diff_pos + k] + old[old_pos + k]) & 0xFF
        new_pos += copy_len
        old_pos += copy_len
        diff_pos += copy_len

        if new_pos + insert_len > new_size:
            raise ValueError("插入长度超出新文件长度")
        if extra_pos + insert_len > len(extra):
            raise ValueError("EXTRA 块数据不足")
        out[new_pos:new_pos + insert_len] = extra[extra_pos:extra_pos + insert_len]
        new_pos += insert_len
        extra_pos += insert_len
        old_pos += seek_len
    return bytes(out)

File: pacc-cross-platform-updater/src/main/java/com/potatotv/pcu/BsPatch.java
package com.potatotv.pcu;

import java.util.Arrays;

/**
 * bsdiff 应用端：把 {@code BSDIFF40} 补丁打到旧文件上，还原出新文件。
 *
 * <p>客户端只做应用，不做生成。补丁可能来自服务端的其它语言实现（例如 python 的
 * {@code bz2}），因此这里只按格式解析，不假设生成端是谁：魔数、三个长度、三个 bzip2
 * 子流都当外部输入校验。</p>
 *
 * <p>校验一律从严：魔数不对、长度是负数或超出实际可用数据、CTRL 里的三元组不够、
 * 复制区间越过旧文件、插入区间越过 EXTRA 块，全部抛 {@link PcuException}。
 * 不做静默截断，也不给缺失的字节补零——补丁被截断或篡改时，宁可失败也不能产出一个
 * 「看起来能装但内容是错的」制品，那会把问题推迟到运行期才炸。</p>
 */
public final class BsPatch {

    private static final byte[] MAGIC = {'B', 'S', 'D', 'I', 'F', 'F', '4', '0'};
    private static final int HEADER_SIZE = 32;
    private static final int CTRL_TUPLE_SIZE = 24;

    private BsPatch() {
    }

    /**
     * 用补丁把旧文件还原成新文件。
     *
     * @param oldData   当前版本的完整文件字节
     * @param patchData BSDIFF40 补丁字节
     */
    public static byte[] patch(byte[] oldData, byte[] patchData) {
        if (oldData == null || patchData == null) {
            throw new PcuException("bsdiff 应用的输入不能为 null");
        }
        if (patchData.length < HEADER_SIZE) {
            throw new PcuException("差分补丁长度不足，无法容纳 BSDIFF40 头");
        }
        for (int i = 0; i < MAGIC.length; i++) {
            if (patchData[i] != MAGIC[i]) {
                throw new PcuException("差分补丁魔数不是 BSDIFF40");
            }
        }

        long ctrlLen = BsDiff.readOffset(patchData, 8);
        long diffLen = BsDiff.readOffset(patchData, 16);
        long newSize = BsDiff.readOffset(patchData, 24);
        if (ctrlLen < 0) {
            throw new PcuException("差分补丁的 CTRL 块长度为负：" + ctrlLen);
        }
        if (diffLen < 0) {
            throw new PcuException("差分补丁的 DIFF 块长度为负：" + diffLen);
        }
        if (newSize < 0) {
            throw new PcuException("差分补丁声明的新文件长度为负：" + newSize);
        }
        if (newSize > Integer.MAX_VALUE - 8L) {
            throw new PcuException("差分补丁声明的新文件过大：" + newSize);
        }
        // 用 long 比较，避免 ctrlLen + diffLen 在 int 上回绕成负数绕过检查。
        long payload = ctrlLen + diffLen;
        if (payload > patchData.length - HEADER_SIZE || payload < 0) {
            throw new PcuException("差分补丁声明的 CTRL/DIFF 块长度超出实际数据"
                    + "（ctrl=" + ctrlLen + " diff=" + diffLen
                    + " 可用=" + (patchData.length - HEADER_SIZE) + "）");
        }

        int ctrlOff = HEADER_SIZE;
        int diffOff = (int) (HEADER_SIZE + ctrlLen);
        int extraOff = (int) (HEADER_SIZE + ctrlLen + diffLen);

        // 三个块各自带输出上限，别让「解压出来到底多大」由补丁自己说了算：
        // DIFF 块只提供 copyLen 字节、EXTRA 块只提供 insertLen 字节，而每次循环都会把
        // copyLen + insertLen 记进 newPos 且 newPos 不超过 newSize，所以这两个块的实际
        // 长度必然不超过 newSize。超过就说明补丁是构造出来的，直接失败，不要先解压再判断。
        long blockLimit = Math.min(newSize, BZip2.MAX_DECOMPRESS_BYTES);
        byte[] ctrl = BZip2.decompress(Arrays.copyOfRange(patchData, ctrlOff, diffOff));
        byte[] diff = BZip2.decompress(Arrays.copyOfRange(patchData, diffOff, extraOff), blockLimit);
        byte[] extra = BZip2.decompress(Arrays.copyOfRange(patchData, extraOff, patchData.length), blockLimit);

        // 分配必须先于「声明的 newSize」被证伪，而不是先分配再校验：一个 32 字节的伪造补丁
        // 把 newSize 写成 2GB，就能让端侧直接 OOM（Error 不受 catch(RuntimeException) 约束）。
        // 可证伪的依据：CTRL 里每个三元组的 copyLen + insertLen 之和恰好等于 newSize（循环
        // 退出条件就是 newPos == newSize），而 copyLen 只从 DIFF 取字节、insertLen 只从 EXTRA
        // 取字节，所以 newSize 不可能超过这两个解压块的长度之和。
        if (newSize > (long) diff.length + extra.length) {
            throw new PcuException("差分补丁声明的新文件长度超出 DIFF/EXTRA 块可用数据"
                    + "（new=" + newSize + " diff=" + diff.length + " extra=" + extra.length + "）");
        }
        byte[] out = new byte[(int) newSize];
        long oldPos = 0;
        int newPos = 0;
        int ctrlPos = 0;
        int diffPos = 0;
        int extraPos = 0;

        while (newPos < newSize) {
            // 三元组必须完整；这条检查同时也是死循环的出口
            if (ctrlPos + CTRL_TUPLE_SIZE > ctrl.length) {
                throw new PcuException("差分补丁的 CTRL 块不完整，缺少三元组");
            }
            long copyLen = BsDiff.readOffset(ctrl, ctrlPos);
            long insertLen = BsDiff.readOffset(ctrl, ctrlPos + 8);
            long seekLen = BsDiff.readOffset(ctrl, ctrlPos + 16);
            ctrlPos += CTRL_TUPLE_SIZE;
            if (copyLen < 0) {
                throw new PcuException("差分补丁的复制长度为负：" + copyLen);
            }
            if (insertLen < 0) {
                throw new PcuException("差分补丁的插入长度为负：" + insertLen);
            }
            if (newPos + copyLen > newSize) {
                throw new PcuException("差分补丁的复制长度超出新文件长度");
            }
            if (copyLen > diff.length - diffPos) {
                throw new PcuException("差分补丁的 DIFF 块数据不足");
            }
            if (copyLen > 0) {
                if (oldPos < 0 || oldPos + copyLen > oldData.length) {
                    throw new PcuException("差分补丁的复制区间超出旧文件范围");
                }
                for (int i = 0; i < copyLen; i++) {
                    out[newPos + i] = (byte) (diff[diffPos + i] + oldData[(int) (oldPos + i)]);
                }
            }
            newPos += (int) copyLen;
            oldPos += copyLen;
            diffPos += (int) copyLen;

            if (newPos + insertLen > newSize) {
                throw new PcuException("差分补丁的插入长度超出新文件长度");
            }
            if (insertLen > extra.length - extraPos) {
                throw new PcuException("差分补丁的 EXTRA 块数据不足");
            }
            if (insertLen > 0) {
                System.arraycopy(extra, extraPos, out, newPos, (int) insertLen);
            }
            newPos += (int) insertLen;
            extraPos += (int) insertLen;

            // seek 可以为负：补丁就是靠它回到旧文件里更早的位置再复制
            oldPos += seekLen;
        }
        return out;
    }
}